File: Hyper_Drive/place_in_opt/pleora/ebus_sdk/Ubuntu-22.04-x86_64/share/samples/SoftDeviceGEV/MyRegisterEventSink.cpp
// *****************************************************************************
//
// Copyright (c) 2018, Pleora Technologies Inc., All rights reserved.
//
// *****************************************************************************

#include "Defines.h"
#include "MyRegisterEventSink.h"


///
/// \brief Constructor
///

MyRegisterEventSink::MyRegisterEventSink()
{
    mSampleEnum = 0;
    mSampleStrings.assign( SAMPLEENUMCOUNT, "" );
    mSampleBooleans.assign( SAMPLEENUMCOUNT, 0 );
}


///
/// \brief Destructor
///

MyRegisterEventSink::~MyRegisterEventSink()
{

}


///
/// \brief Pre-read notification - usually a good place to update the register content
///

PvResult MyRegisterEventSink::PreRead( IPvRegister *aRegister )
{
    std::cout << aRegister->GetName().GetAscii() << " PreRead" << std::endl;
    if ( aRegister->GetAddress() == SAMPLESTRINGADDR )
    {
        aRegister->Write( mSampleStrings[ mSampleEnum ] );
    }
    if ( aRegister->GetAddress() == SAMPLEBOOLEANADDR )
    {
        aRegister->Write( mSampleBooleans[ mSampleEnum ] );
    }
    return PvResult::Code::OK;
}


///
/// \brief Post-read nofitication
///

void MyRegisterEventSink::PostRead( IPvRegister *aRegister )
{
    std::cout << aRegister->GetName().GetAscii() << " PostRead" << std::endl;
}


///
/// \brief Pre-write notification - this is where a new register value is usually validated
///

PvResult MyRegisterEventSink::PreWrite( IPvRegister *aRegister )
{
    std::cout << aRegister->GetName().GetAscii() << " PreWrite" << std::endl;
    return PvResult::Code::OK;
}


///
/// \brief Post-write notification: react to a register write
///

void MyRegisterEventSink::PostWrite( IPvRegister *aRegister )
{
    // We need to reset command registers to 0 after activation for IsDone
    if ( aRegister->GetAddress() == SAMPLECOMMANDADDR )
    {
        mSampleStrings[ mSampleEnum ] = "";
        aRegister->Write( 0 );
    }
    if ( aRegister->GetAddress() == SAMPLEENUMADDR )
    {
        aRegister->Read( mSampleEnum );
    }
    if ( aRegister->GetAddress() == SAMPLESTRINGADDR )
    {
        aRegister->Read( mSampleStrings[ mSampleEnum ] );
    }
    if ( aRegister->GetAddress() == SAMPLEBOOLEANADDR )
    {
        aRegister->Read( mSampleBooleans[ mSampleEnum ] );
    }

    std::cout << aRegister->GetName().GetAscii() << " PostWrite" << std::endl;
}


///
/// \brief Manage Userset persistence for nodes which vary by a selector
///
PvResult MyRegisterEventSink::Persist( IPvRegister *aRegister, IPvRegisterStore *aStore )
{
    switch ( aRegister->GetAddress() )
    {
    case SAMPLESTRINGADDR:
    case SAMPLEBOOLEANADDR:
        return PersistSampleEnum( aRegister, aStore );
    }
    return PvResult::Code::NOT_IMPLEMENTED;
}


//
/// \brief Manage Userset persistence for nodes which vary by SampleEnum
///
/// Since these nodes have a single address we need to explicitly save/load all
/// possible values over the selector range
///
PvResult MyRegisterEventSink::PersistSampleEnum( IPvRegister *aRegister, IPvRegisterStore *aStore ) 
{
    // Preserve current mSampleEnum value
    uint32_t lEnumOrg = mSampleEnum;

    for ( int i = 0; i < SAMPLEENUMCOUNT; i++ )
    {
       // Select Sample Enum
       mSampleEnum = i;

       // Create persistence suffix
       long long lEnumSuffix = i + 1;
       std::string lSuffix = std::string("EnumEntry") + std::to_string(lEnumSuffix) ;

       // Persist selected register (save or load depending on context)
       aStore->Persist( aRegister, lSuffix.c_str() );
    }

    mSampleEnum = lEnumOrg;
    return PvResult::Code::OK;
 
}

File: Hyper_Drive/place_in_opt/pleora/ebus_sdk/Ubuntu-22.04-x86_64/share/samples/python/ebus/SoftDeviceGEV/MyRegisterEventSink.py
#!/usr/bin/env python3

'''
*****************************************************************************
*
*   Copyright (c) 2020, Pleora Technologies Inc., All rights reserved.
*
*****************************************************************************
'''

import eBUS as eb
from Defines import *


class MyRegisterEventSink(eb.IPvRegisterEventSink):
    def __init__(self):
        super().__init__()
        self._sample_enum = 0
        self._sample_strings = [ "", "", "" ]
        self._sample_booleans = [ False, False, False ]

    def PreRead(self, register):
        print(f"{register.GetName()} PreRead")
        if register.GetAddress() == SAMPLESTRINGADDR:
            register.Write( self._sample_strings[ self._sample_enum ] )
        if register.GetAddress() == SAMPLEBOOLEANADDR:
            register.Write( self._sample_booleans[ self._sample_enum ] )
        return eb.PV_OK

    def PostRead(self, register):
        print(f"{register.GetName()} PostRead")

    def PreWrite(self, register):
        print(f"{register.GetName()} PreWrite")
        return eb.PV_OK

    def PostWrite(self, register):
        # We need to reset command registers to 0 after activation for IsDone
        if register.GetAddress() == SAMPLECOMMANDADDR:
            self._sample_strings[self._sample_enum] = ""
            register.Write(0)
        if register.GetAddress() == SAMPLEENUMADDR:
            result, self._sample_enum = register.ReadInt()
        if register.GetAddress() == SAMPLESTRINGADDR:
            result, self._sample_strings[self._sample_enum] = register.ReadString()
        if register.GetAddress() == SAMPLEBOOLEANADDR:
            result, self._sample_booleans[self._sample_enum] = register.ReadInt()

        print(f"{register.GetName()} PostWrite")


    def Persist(self, register, store):
        # Manage Userset persistence for nodes which vary by a selector
        if register.GetAddress() == SAMPLESTRINGADDR:
           return self.PersistSampleEnum( register, store )
        if register.GetAddress() == SAMPLEBOOLEANADDR:
           return self.PersistSampleEnum( register, store )
        return eb.PV_NOT_IMPLEMENTED;


    def PersistSampleEnum(self, register, store):
        # Manage Userset persistence for nodes which vary by a SampleEnum
        # Preserve current mSampleEnum value
        enum_org = self._sample_enum

        for i in range(SAMPLEENUMCOUNT):
            # Select Sample Enum
            self._sample_enum = i

            # Create persistence suffix
            suffix = "EnumEntry" + str(i + 1)

            # Persist selected register (save or load depending on context)
            store.Persist(register, suffix)

        self._sample_enum = enum_org;
        return eb.PV_OK;
